[OBX-UX-MNGMT] Allow decimals in the threshold filed for the Failed transaction rate threshold rule

fkanout opened the pull request:
## Summary

It fixes #184527 by allowing decimals in the threshold value. This was a purely UI restriction.

![Screenshot 2024-06-03 at 14 29 23](https://github.com/elastic/kibana/assets/6838659/cce7eb52-977d-403d-b4e3-6e9e5a87a8d3)

### Release note:
Now we allow the decimals value in the threshold filed for the Failed transaction rate threshold rule e.g. `30.50`% or `0.1` %

fkanout committed:
Allow decimals for the threshold value.

elasticmachine commented:
Pinging @elastic/obs-ux-management-team (Team:obs-ux-management)

elasticmachine commented:
Pinging @elastic/obs-ux-infra_services-team (Team:obs-ux-infra_services)

apmmachine commented:
### :robot: GitHub comments

<details><summary>Expand to view the GitHub comments</summary>
<p>

Just comment with:
- `/oblt-deploy` : Deploy a Kibana instance using the Observability test environments.
- `run` `docs-build` : Re-trigger the docs validation. (use unformatted text in the comment!)
</p>
</details>

fkanout commented on x-pack/plugins/observability_solution/apm/public/components/alerting/utils/fields.tsx:
```diff
@@ -42,7 +42,7 @@ export function ServiceField({
         customOptions={allowAll ? [{ label: allOptionText, value: undefined }] : undefined}
         customOptionText={i18n.translate('xpack.apm.serviceNamesSelectCustomOptionText', {
-          defaultMessage: `Add '{searchValue}' as a new service name`,
+          defaultMessage: 'Add \\{searchValue\\} as a new service name',
```
🤷🏼

zwang1-godaddy commented:
Hey Kibana, glad to see this feature is almost ready to go. Do you have a time estimate for when it will be released? Looking forward to seeing it in production.

jennypavlova reviewed (commented):
The rule is working as expected. I have a question regarding the input - The supported separator is currently coma (`,`) and the dot (`.`) is not fully supported: is that intentional?
When I tried to put a `5.5` pressing the `.` resets the value to 0 (If I type 33 and put the `.` between the digits it works and it shows `3.3`) So it resets to 0 only when the `.` is at the end of the input while typing:

https://github.com/elastic/kibana/assets/14139027/826db71c-9727-445b-a8a7-5167f758bcd0

With `,` as separator it looks good:

<img width="596" alt="image" src="https://github.com/elastic/kibana/assets/14139027/b1338b8b-7dd4-426d-80a1-fa0880cb005d">

To reproduce try typing a number like `5` then `.` then `5` (It resets to `0` after typing `.`) In case `.` is not a valid separator it works

fkanout commented:
> The rule is working as expected. I have a question regarding the input - The supported separator is currently coma (`,`) and the dot (`.`) is not fully supported: is that intentional? When I tried to put a `5.5` pressing the `.` resets the value to 0 (If I type 33 and put the `.` between the digits it works and it shows `3.3`) So it resets to 0 only when the `.` is at the end of the input while typing:
> 
>  Screen.Recording.2024-06-06.at.14.28.51.mov 
> With `,` as separator it looks good:
> 
> <img alt="image" width="596" src="https://private-user-images.githubusercontent.com/14139027/337262436-b1338b8b-7dd4-426d-80a1-fa0880cb005d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTgxODA5MDUsIm5iZiI6MTcxODE4MDYwNSwicGF0aCI6Ii8xNDEzOTAyNy8zMzcyNjI0MzYtYjEzMzhiOGItN2RkNC00MjZkLTgwYTEtZmEwODgwY2IwMDVkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjEyVDA4MjMyNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTk3ZmMwMDdlZjczZDcyYjgyYjQ4ZTk2OTNhN2NkMzBjYzliNjI3NjEyMWIyOWUwMGNjOTVhMzFmODY3ZTc4YjUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.SPFkG-a0QLyJrzRlkpbyqaaB54CB_YHjCj3Y8sfetBk">
> To reproduce try typing a number like `5` then `.` then `5` (It resets to `0` after typing `.`) In case `.` is not a valid separator it works

Good catch @jennypavlova. I'm checking that!

fkanout committed:
Merge branch 'main' into 184527-allow-decimals-failed-transaction-rate-rule-threshold

fkanout committed:
Merge branch 'main' into 184527-allow-decimals-failed-transaction-rate-rule-threshold

fkanout committed:
Merge branch 'main' into 184527-allow-decimals-failed-transaction-rate-rule-threshold

fkanout committed:
Use better validation and fix the decimal separator

fkanout commented:
@zwang1-godaddy, sorry about that delay. I'm currently working half-days, and the priority was shifting, but the good new we could still ship it in `8.15`.

fkanout commented:
> The rule is working as expected. I have a question regarding the input - The supported separator is currently coma (`,`) and the dot (`.`) is not fully supported: is that intentional? When I tried to put a `5.5` pressing the `.` resets the value to 0 (If I type 33 and put the `.` between the digits it works and it shows `3.3`) So it resets to 0 only when the `.` is at the end of the input while typing:
> 
>  Screen.Recording.2024-06-06.at.14.28.51.mov 
> With `,` as separator it looks good:
> 
> <img alt="image" width="596" src="https://private-user-images.githubusercontent.com/14139027/337262436-b1338b8b-7dd4-426d-80a1-fa0880cb005d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTg5MDExOTgsIm5iZiI6MTcxODkwMDg5OCwicGF0aCI6Ii8xNDEzOTAyNy8zMzcyNjI0MzYtYjEzMzhiOGItN2RkNC00MjZkLTgwYTEtZmEwODgwY2IwMDVkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjIwVDE2MjgxOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTc3NjAwNGRmMjlkM2FiNmI1ODBkMzYxNmI2MzFkZTMwYjE1MzRlYTgwYzlhY2Q3MzE3MzI3YjRlMmJjMmY2NTMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.QiebW7UnvISATpt8CZu53AgFo_mguxuhJAIDn83ql8A">
> To reproduce try typing a number like `5` then `.` then `5` (It resets to `0` after typing `.`) In case `.` is not a valid separator it works

@jennypavlova, I fixed this issue and enhanced the threshold value validation and component validation. Please give it a look and let me know if anything!

fkanout committed:
Merge branch 'main' into 184527-allow-decimals-failed-transaction-rate-rule-threshold

jennypavlova commented on x-pack/plugins/observability_solution/apm/public/components/alerting/utils/fields.tsx:
```diff
@@ -203,22 +203,50 @@ export function IsAboveField({
-      })}
+    <EuiPopover
+      isOpen={thresholdPopoverOpen}
+      anchorPosition={'downLeft'}
```
nit
```suggestion
      anchorPosition="downLeft"
```

fkanout replied:
Done

jennypavlova commented on x-pack/plugins/observability_solution/apm/public/components/alerting/utils/fields.tsx:
```diff
@@ -203,22 +203,50 @@ export function IsAboveField({
 }) {
+  const [thresholdPopoverOpen, serThresholdPopoverOpen] = useState(false);
+  const [isAboveValue, setIsAboveValue] = useState(String(value));
+  const isInvalid = isNaN(Number(isAboveValue));
```
nit: This can be a function so you can pass the value -  something like `const isValidNumber = (value: number): boolean => !isNaN(Number(value))` and reuse it on lines 241 (`isValidNumber(thresholdVal)`) and 246 (`!isValidNumber(isAboveValue)`)

fkanout replied:
Sure

jennypavlova reviewed (approved):
Thanks for fixing the issue! It works fine with both (`.` and `,`) separators now 💯 

I added some nits and one small nit on the UI side (not blocking but I think it's good to have): we can add an error message like in the other rules so we have the icon now: 
![image](https://github.com/elastic/kibana/assets/14139027/e5cc1974-cce2-4e19-af5e-90bf6cbc174b)

we can also have an error message like in the inventory rule: 
![image](https://github.com/elastic/kibana/assets/14139027/49850384-2323-44ac-972c-3fda70e0a3e1)

wdyt?

fkanout committed:
code review

fkanout commented:
> Thanks for fixing the issue! It works fine with both (`.` and `,`) separators now 💯
> 
> I added some nits and one small nit on the UI side (not blocking but I think it's good to have): we can add an error message like in the other rules so we have the icon now: ![image](https://private-user-images.githubusercontent.com/14139027/341504633-e5cc1974-cce2-4e19-af5e-90bf6cbc174b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTkyMjk5NzYsIm5iZiI6MTcxOTIyOTY3NiwicGF0aCI6Ii8xNDEzOTAyNy8zNDE1MDQ2MzMtZTVjYzE5NzQtY2NlMi00ZTE5LWFmNWUtOTBiZjZjYmMxNzRiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjI0VDExNDc1NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTdmZDMzYmE5YjA4NjkzMmU5NjEwZjA1Y2FlZTQ4NDA0OGE0OGZlZjI1N2Y5ZDY2ZjI2MGY2OTVhZTEyZjRjNzUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.-sQKv8gL7uvWlVYWXM_H4OsEyY-C4K9mtdNUhOLwDwE)
> 
> we can also have an error message like in the inventory rule: ![image](https://private-user-images.githubusercontent.com/14139027/341504703-49850384-2323-44ac-972c-3fda70e0a3e1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTkyMjk5NzYsIm5iZiI6MTcxOTIyOTY3NiwicGF0aCI6Ii8xNDEzOTAyNy8zNDE1MDQ3MDMtNDk4NTAzODQtMjMyMy00NGFjLTk3MmMtM2ZkYTcwZTBhM2UxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjI0VDExNDc1NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWFjZWYxZjFkYjMyMTdlYTU5MTNhODY2ZWY1MzE3N2U1YmVhZjU2NzI0ZDJjOGJmNjY0YzNjM2NjY2Q3ZWRmOGEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.W8jw5kuF_4ThH-CCjSK67Erw_LEB5CRakAViQ_EwBxY)
> 
> wdyt?

I agree. I added it

fkanout committed:
Merge branch 'main' into 184527-allow-decimals-failed-transaction-rate-rule-threshold
